### ISSUE TITLE
Cross-platform compatibility

### ISSUE THREAD

**Issue by walkawayy** [LostArtefacts/TRX#716]
- [x] Replace wgl_ext.h with cross platform variant
- [x] Remove HWND and HINSTANCE usages
- [x] Rename JSON paths to use `/` instead of `\\` separators
- [x] Case insensitive directory and file detection
- [x] Determine how to package FFmpeg and SDL2
- [x] Test on Linux
- [x] Test on MacOS

---
Looking at the README's [road map](https://github.com/rr-/Tomb1Main#road-map), it seems like we are moving towards being able to support cross-platform builds. @carlo-bramini are both of the following tasks done once #714 merges?

```
Replace wgl_ext.h with cross platform variant
Remove HWND and HINSTANCE usages
```

Do you have a list of what else needs to be done that we could track in this issue? Also it seems like you've been focusing on Linux, but is MacOS compatibility a possibility too?

**Comment by carlo-bramini** [comment]
About these points:

```
Replace wgl_ext.h with cross platform variant
Remove HWND and HINSTANCE usages
```

I have compiled and executed the engine also on linux, so it is technically able to work, although the .json5 files still need to be adjusted for fixing directory separators from `'\\'` to `'/'` and some decisions should be made since the filesystem is case sensitve on other platforms.

At the moment, I compiled and tested the engine with the demo level and some hacks with these platforms:
- i686, 32-bit executable, on Windows 11, Windows 10, Windows 7 and Windows XP.
- x86_64, 64 bit executable, on Windows 11, Windows 10 and Windows 7.
- x86_64, 64 bit executable, on Linux Debian.
- ARM64, 64 bit executable, on Linux Raspberry Pi OS (needs an hack to the shaders, see #685).

I compiled my builds with CYGWIN and MinGW cross compilers. However, CYGWIN does not provide FFmpeg and the current version of SDL2 is too old for compiling Tomb1Main, but here I made public the packages made by me for CYGWIN:

https://github.com/carlo-bramini/packages-cygwin

I also compiled Tomb1Main by using the POSIX layer of CYGWIN and I run the engine with the CYGWIN/X server, but actually it was just the development tool for running it on UNIX-like systems.

So, in my opinion, those two points seems to be resolved.

About MacOS, unfortunately I cannot test if it works because I have no way to run it.

**Comment by walkawayy** [comment]
Ok cool. I think the goal should be to remove the need for any hacks if we can. (Except it seems the demo support will require some, but that's a whole different issue IMO.)

From what I read, I think we could safely rename all JSON paths to use `/` like you said. That should work on both Windows and Linux.

Linux is case sensitive. So, I think we need to determine what filenames Steam, GOG, game discs, etc ship with and name the defaults in the JSON based on what is most common. Maybe we can try to fall back to all lower / upper case names if the default filename read fails. And maybe even adjust the crash warning if no level / FMV files can be read.

As for the packages, I think @rr- did something similar to get FFmpeg and SDL2 in the Docker containers, but I don't know that for sure. I tagged rr, so we can get some clarification.

I have a Mac to test with but it's an Intel processor and not a new M1/M2.

I'll update this issue to track the remaining tasks.

**Comment by rr-** [comment]
We have a few options:

1. ship a static binary blob that has everything and is big. No issues with ABI. I'd prefer to do it in a Linux Docker container.
2. ship a dynamically linked binary blob. Requires the user to install small binary blob packages such as ffmpeg and sdl2. Compiling against a version that's too old may cause crashes and ABI issues.
3. let people compile the game themselves. Requires the user to install big development packages such as ffmpeg-dev and sdl2-dev. As long as our build system is relatively sane and I believe it is, I don't think we'll have to do much hand-holding here.

> Maybe we can try to fall back to all lower / upper case names if the default filename read fails.

`File_Open` needs to be made case insensitive in my opinion on Linux. This can be made by enumerating files in the parent directory (or directories!!) if they do not exist.

**Comment by carlo-bramini** [comment]
> `File_Open` needs to be made case insensitive in my opinion on Linux. This can be made by enumerating files in the parent directory (or directories!!) if they do not exist.

I was thinking the same thing, by doing a search and adjust the name of the file to the first one that matches to the target.
BTW, Windows can be also case sensitive on user's request, so I would make this change for every platform.

**Comment by carlo-bramini** [comment]
I compiled and tested successfully the engine with the demo level also on Haiku/BeOS.
I had to do two quick fixes to the code for reaching this result:

(1)
src/gfx/gl/gl_core_3_3.c:
I removed all platform dependant `#ifdef...#endif` and I replaced it with:

```
#include "SDL2/SDL.h"
#define IntGetProcAddress(_name) SDL_GL_GetProcAddress(_name)
```

(2)
The SDL2 library doesn't support `SDL_CONTROLLER_TYPE_NINTENDO_SWITCH_JOYCON_PAIR`, which seems to be a quite new addition.
So, I added an `#if SDL_VERSION_ATLEAST(2,24,0)...#endif` around that identifier.

After that, you can see this image on your screen:

![VirtualBox_Haiku 64_23_01_2023_12_25_54](https://user-images.githubusercontent.com/30959007/214033691-44144c2a-150d-4739-a422-794ac6716524.png)

Perhaps, it would be worth to evaluate if (1) is good on all supported platforms.
On Windows and Linux, it is working fine.
About the Macs, unfortunately, I do not know, but since we are relying on SDL, there are high chances that it will be also working fine.

NOTE: this screenshot have some text in Italian because I translated the string into the .json5 file in my language.

**Comment by rr-** [comment]
I'm all for removing `src/gfx/gl/gl_core_3_3.c` in favor of SDL if possible.

**Comment by carlo-bramini** [comment]
I'm wondering if supporting also SDL-2.0.x and SDL-1.2.x could be accepted as fixes for increasing cross platform compatibility.

**Comment by walkawayy** [comment]
> I'm wondering if supporting also SDL-2.0.x and SDL-1.2.x could be accepted as fixes for increasing cross platform compatibility.

I don't think newer SDL versions require specific hardware, so I don't see why we need to. Trying to support things like older OpenGL shaders can help people run the game. But I don't think adding support for old SDL versions is needed. And the SDL version is set in the Docker image, so there normally shouldn't be any reason to mess with it.

**Comment by legluondunet** [comment]
Hello, about Cross-platform compatibility, do you plan to develop a config tool and an installer more portable (without dotnet).

**Comment by walkawayy** [comment]
> Hello, about Cross-platform compatibility, do you plan to develop a config tool and an installer more portable (without dotnet).

I don't think anything has been decided regarding a Linux installer or config tool. The plan is to first officially support Linux compilation and go from there.

**Comment by vvs-** [comment]
Compiled it today on NixOS using CMake. There were errors from gcc 10.3.0 `-Werror=format-security`, missing symbols from SDL2 2.0.14 `SDL_CONTROLLER_TYPE_NINTENDO_SWITCH_JOYCON_PAIR` and it needed a few additional lines stemming from recent additions of injection framework. After dealing with those it runs without a crash even on OpenGL 2.1 with a slightly massaged shaders. The only strange thing is that it reports that `ARB_explicit_attrib_location` wasn't found though actually it's there.

That's not a complain, but just facts to consider.

**Comment by walkawayy** [comment]
Case insensitive directory and file detection merged in #846.

**Comment by rr-** [comment]
Linux support has landed on develop; discussion on Macs support shall continue in #997.
